### ISSUE TITLE
NSCBC implement in openfoam-9

### ISSUE THREAD

**Issue by jiaqiwang969** [synthetik-technologies/blastfoam#55]
Hi, I am Jiaqi, a phd student in aeroacoustic.

I try a lot repository of compressible foam, finally I choose blastfoam for my dlr-buffet project. 

But, the boundary condition now is only limited in waveTransimision, which is not really good for inter-flow simulation (with terrible dirft). I have developed [NSCBC](https://github.com/jiaqiwang969/openfoam-docker) to solve it. However, I am only test it in Openfoam-v2006,v2012.

 Now, I am troubled with implement it  in openfoam-9. Could you help me to debug?

**Comment by jheylmun** [comment]
Hi Jiaqi,
I would be happy to. One thing that may help would for the initial port would be to use blastReactingFoam instead of blastFoam since the way in which the speed of sound would be calculated is different between the standard OpenFOAM thermo and blastFoam's.

Thanks,
Jeff

**Comment by jiaqiwang969** [comment]
Right, I found this diffenrence when I try to use thermo:psi. For BlastFoam is not avaiable.  May I ask what's the different between blastReactingFoam and blastFoam.   Right now, my another prohect "DLR-buffet", using blastFoam, the turbulence model is include by adding fvModel. I am a little confuse about how it works.

**Comment by jheylmun** [comment]
Really the only difference between the solvers is that blastReactingFoam uses the standard OpenFOAM thermo classes, just like rhoCentralFoam, but also has ability to use the ODE time integration, flux schemes, and other numerical tools that rhoCentralFoam does not. Reactions and combustion models are also included.

What do you mean it is "included by adding fvModels"? As long as you have a non-zero viscosity specified in phaseProperties (blastFoam) or thermophysicalProperties (blastReactingFoam), you should be required to provide either turbulenceProperties or momentumTransportProperties.

**Comment by jheylmun** [comment]
One other thing I just noticed in the implementation that will make using your boundary conditions with more difficult with blastFoam is that you are using the ddt scheme to calculate the boundary which is not actually used. Instead there are ODE coefficients that are used to blend both the starting values as well as the changes in the advancing variables. This results in the calculation of the refValue and refGrad values not being consistent.

**Comment by jiaqiwang969** [comment]
> Really the only difference between the solvers is that blastReactingFoam uses the standard OpenFOAM thermo classes, just like rhoCentralFoam, but also has ability to use the ODE time integration, flux schemes, and other numerical tools that rhoCentralFoam does not. Reactions and combustion models are also included.
> 
> What do you mean it is "included by adding fvModels"? As long as you have a non-zero viscosity specified in phaseProperties (blastFoam) or thermophysicalProperties (blastReactingFoam), you should be required to provide either turbulenceProperties or momentumTransportProperties.

But It seems the blastReactingFoam solver is more unstable than blastFoam or rhoCentralFoam.
I test the same case, with those three solver. blastReactingFoam is always broken up.

**Comment by jheylmun** [comment]
Could you provide the case you are using? The set up may be slightly different between the three

**Comment by jiaqiwang969** [comment]
Sorry for the late reply，I haved upload three cases in https://github.com/jiaqiwang969/ExampleCase-blastFoam

**Comment by jiaqiwang969** [comment]
For this setting，blastFoam & rhoCentralFoam is running well. blastReactingFoam is broken up.

While the results between blastFoam and rhoCentralFoam is different, when both with Kurganov 
 fluxSchme and same parameter setting. rhoCentralFoam sound like more dissipative than blastFoam. rhoCentralFoam is more close to experiments.

**Comment by jheylmun** [comment]
The was unable to run the cases you provided, but one thing that I noticed was the schemes you used varied between the solvers. The quadraticMUSCL scheme used in the blastFoam and blastReactingFoam cases will tend to be less dissipative that the vanLeer schemes used in the rhoCentralFoam case. I would also recommend using the Euler timeIntegrationScheme for consistency. While it would probably be worth while to compare Euler and higher order time integration schemes in blastFoam, I would suggest using the same schemes to compare to rhoCentralFoam.
